### PR TITLE
Renew benchmark

### PR DESCRIPTION
**Pull request by hhorii**
### Summary

Renew asv benchmark for continuous performance evaluation

### Details and comments

This PR aims to renew asv benchmark for qiskit-aer.

In addition, this PR allows users to check performance with shell. 

```
cd test/
PYTHONPATH=./ python -u benchmark2/particular_quantum_circuits.py
```

- [x] Evaluate most of circuits in Qiskit Circuit Library
- [x] Evaluate noise simulation
- [x] Evaluate sampling and expectation value calculation
- [x] Support Linux environment
- [x] Support mac environment
- [x] Evaluate chemistry application
- [x] Add command line arguments to specify benchmark configuration

### CODE REVIEW

**Comment by hhorii** [comment]
Sampled results are here: https://hhorii.github.io/qiskit-aer/html/

File: test/benchmark/basic.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    #    self._run(self.RUNTIME_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_extended_stabilizer(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_EXTENDED_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_unitary_matrix(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_CPU, app, qubit)
+        
+    #def time_unitary_matrix_gpu(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_GPU, app, qubit)
```
Comment (vvilpas) [inline test/benchmark/basic.py:62]: Remove?

File: test/benchmark/noise.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    #    self._run(self.RUNTIME_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_extended_stabilizer(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_EXTENDED_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_unitary_matrix(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_CPU, app, qubit)
+        
+    #def time_unitary_matrix_gpu(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_GPU, app, qubit)
```
Comment (vvilpas) [inline test/benchmark/noise.py:62]: Remove?

File: test/benchmark/output.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    #    self._run(self.RUNTIME_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_extended_stabilizer(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_EXTENDED_STABILIZER_CPU, app, measure, measure_count, noise_name, qubit)
+        
+    #def time_unitary_matrix(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_CPU, app, qubit)
+        
+    #def time_unitary_matrix_gpu(self, app, measure, measure_count, noise_name, qubit):
+    #    self._run(self.RUNTIME_UNITARY_MATRIX_GPU, app, qubit)
```
Comment (vvilpas) [inline test/benchmark/output.py:62]: Remove?

File: test/benchmark/simulator_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        
+        #if self.RUNTIME_STABILIZER_CPU in runtime_names:
+        #    self.simulators[self.RUNTIME_STABILIZER_CPU] = QASM_SIMULATOR
+        #    self.backend_options_list[self.RUNTIME_STABILIZER_CPU] = { 'method': self.RUNTIME_STABILIZER_CPU }
+        #    self.backend_qubits[self.RUNTIME_STABILIZER_CPU] = self.qubits
+        
+        #if self.RUNTIME_EXTENDED_STABILIZER_CPU in runtime_names:
+        #    self.simulators[self.RUNTIME_EXTENDED_STABILIZER_CPU] = QASM_SIMULATOR
+        #    self.backend_options_list[self.RUNTIME_EXTENDED_STABILIZER_CPU] = { 'method': self.RUNTIME_EXTENDED_STABILIZER_CPU }
+        #    self.backend_qubits[self.RUNTIME_EXTENDED_STABILIZER_CPU] = self.qubits
```
Comment (vvilpas) [inline test/benchmark/simulator_benchmark.py:153]: Remove?

File: test/benchmark/vqe_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            'LiH': ('H .0 .0 .0; Li .0 .0 2.5', 10),                   # qubits: 10
+            'HF': ('H .0 .0 .0; F .0 .0 1.25', 10),                    # qubits: 10
+            #'BeH2': ('H .0 .0 -2.25; Be .0 .0 .0; H .0 .0 2.25', 12),  # qubits: 12
+            #'HCl': 'H .0 .0 .0; Cl .0 .0 2.0',                    # qubits: 18
+            #'Li2': 'Li .0 .0 .0; Li .0 .0 3.0',                   # qubits: 18
+            #'N2': 'Li .0 .0 .0; Li .0 .0 1.8',                    # qubits: 18
+            #'MgH2': 'H .0 .0 -2.3; Mg .0 .0 .0; H .0 .0 2.3',     # qubits: 20
+            #'H2S': 'H .0 .0 -3.4; S .0 .0 .0; H .0 .0 3.4',       # qubits: ...
+            #'Li2S': 'Li .0 .0 -3.0; S .0 .0 .0; Li .0 .0 3.0',    # qubits: ...
+            #'Li2O': 'Li .0 .0 -3.0; O .0 .0 .0; Li .0 .0 3.0',    # qubits: ...
```
Comment (vvilpas) [inline test/benchmark/vqe_benchmark.py:43]: Remove?

**Review by vvilpas — CHANGES_REQUESTED** [review]
Looks good.
A couple of things:
1) Maybe we should split `asv.macos.conf.json` file in regular and cuda one as in linux.
2) Is dir benchmark.bak needed? If not, we can delete it.
3) There are some commented sections in the code (see comments) that I guess we can remove.

File: test/asv.macos.conf.json [hunk 6]
```diff
@@ -30,21 +30,25 @@
...
     ],
     "build_command": [
         "python -mpip install -U scikit-build cython",
-        "python -mpip install qiskit-terra",
-        "VERBOSE=1 python setup.py bdist_wheel --dist-dir={build_cache_dir} -- -- -j"
+        "pip install git+https://github.com/Qiskit/qiskit-terra",
+        "pip install git+https://github.com/Qiskit/qiskit-aqua",
+        "pip install pyscf",
+        "pip install matplotlib",
+        "python setup.py bdist_wheel --dist-dir={build_cache_dir} -- -DCMAKE_CXX_COMPILER=g++ -DAER_THRUST_BACKEND=CUDA -- -j"
```
[the remark this answers is not in the document]
Comment (hhorii) [reply test/asv.macos.conf.json:51]: Right. CUDA is not supported. Also, currently, MacOS is not tested. I would like to take it.

**Review by vvilpas — CHANGES_REQUESTED** [review]
Can we restore `asv.macos.conf.json` in case someone with a mac wants to run the benchmark locally (just remove the `-DAER_THRUST_BACKEND=CUDA`)? Or can we use the `linux` one on macos?
Other than that I think everything is ok.

**Comment by hhorii** [comment]
@vvilpas I confirmed that `asv.linux.conf.json` works on Mac. But I guess that it doesn't work on Windows.

**Comment by vvilpas** [comment]
> @vvilpas I confirmed that `asv.linux.conf.json` works on Mac. But I guess that it doesn't work on Windows.

I think we can add it later in case we need it.

**Review by vvilpas — APPROVED** [review]

**Pull request merged**